File: src/main/go/leetcode/lists/offer/Offer28.go
package offer

/**
 * @author wujingxinit@outlook.com
 * @date 2022/12/29 19:09
 */

func isSymmetric(root *TreeNode) bool {
	if root == nil {
		return true
	}
	var dfs func(node1, node2 *TreeNode) bool
	dfs = func(left, right *TreeNode) bool {
		if left == nil && right != nil {
			return false
		}
		if right == nil && left != nil {
			return false
		}
		if left == nil && right == nil {
			return true
		}
		if left.Val != right.Val {
			return false
		}
		return dfs(left.Right, right.Left) && dfs(left.Left, right.Right)
	}
	return dfs(root.Left, root.Right)
}

File: src/main/java/leetcode/lists/offer/Offer28.java
package leetcode.lists.offer;

import common.TreeNode;

/**
 * @author jingxinwu
 * @date 2021-11-21 7:02 下午
 */
public class Offer28 {

    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        }
        return helper(root.left, root.right);
    }

    public boolean helper(TreeNode L, TreeNode R) {
        if (L == null && R == null) {
            return true;
        }
        if (L == null || R == null || L.val != R.val) {
            return false;
        }
        return helper(L.left, R.right) && helper(L.right, R.left);
    }
}
